crates/lib3h/tests/transport.rs

```diff
@@ -0,0 +1,430 @@
+    Message { from: Url, payload: Vec<u8> },
+    Error(String),
+}
+
+// The Mockernet is a Series-of-Tubes, which is the technical term for the
+// sets of crossbeam channels in the bindings that mockernet shuttles
+// data between.
+// The Mockernet simulates a
+pub struct Tube {
+    sender: crossbeam_channel::Sender<(Url, Vec<u8>)>,
+    #[allow(dead_code)]
+    receiver: crossbeam_channel::Receiver<(Url, Vec<u8>)>,
+}
+impl Tube {
+    pub fn new() -> Self {
+        let (sender, receiver) = crossbeam_channel::unbounded::<(Url, Vec<u8>)>();
+        Tube { sender, receiver }
+    }
+}
+impl Mockernet {
+    pub fn new() -> Self {
+        Mockernet {
+            bindings: HashMap::new(),
+            connections: HashMap::new(),
+            errors: Vec::new(),
+        }
+    }
+
+    /// create a binding to the mockernet, without one you can't send or receive
+    pub fn bind(&mut self, url: Url) -> bool {
+        if self.bindings.contains_key(&url) {
+            false
+        } else {
+            self.bindings.insert(url, Tube::new());
+            true
+        }
+    }
+
+    /// remove a binding, this is should trigger an error event
+    pub fn unbind(&mut self, url: Url) {
+        if self.bindings.contains_key(&url) {
+            self.bindings.remove(&url);
+            self.errors
+                .push((url.clone(), format!("{} has become unbound", &url)));
+        }
+    }
+
+    /// send a message to anyone on the Mockernet
+    pub fn send_to(&mut self, to: Url, from: Url, payload: Vec<u8>) -> Result<(), String> {
+        {
+            let _src = self
+                .bindings
+                .get(&from)
+                .ok_or(format!("{} not bound", from))?;
+            let dst = self.bindings.get(&to).ok_or(format!("{} not bound", to))?;
+            dst.sender
+                .send((from.clone(), payload))
+                .map_err(|e| format!("{}", e))?;
+        }
+        self.connect(from.clone(), to.clone());
+        Ok(())
+    }
+
+    /// check to see, for a given Url, if there are any events waiting
+    pub fn process_for(&mut self, address: Url) -> Result<Vec<MockernetEvent>, String> {
+        let mut events = Vec::new();
+
+        // push any errors for this url into the events
+        let errors: Vec<_> = self.errors.drain(0..).collect();
+        for (url, err) in &errors {
+            if url == &address {
+                events.push(MockernetEvent::Error(err.into()));
+            } else {
+                self.errors.push((url.clone(), err.clone()))
+            }
+        }
+
+        let maybe_binding = self.bindings.get(&address);
+        if maybe_binding.is_none() {
+            if errors.len() == 0 {
+                return Err(format!("{} not bound", address));
+            }
+            return Ok(events);
+        }
+        let (from, payload) = maybe_binding
+            .unwrap()
+            .receiver
+            .try_recv()
+            .map_err(|e| format!("{:?}", e))?;
```

Comment:
Is this correct? This may work fine for the deterministic process calls below, but if we integrated @struktured 's process framework, perhaps here we'd want to loop until we get an `Err(Empty)` response, and not report errors for `Err(Empty)`?

Reply:
good call fixed.